library/core/src/iter/adapters/zip.rs

```diff
@@ -178,25 +178,23 @@ macro_rules! zip_impl_general_defaults {
             A: DoubleEndedIterator + ExactSizeIterator,
             B: DoubleEndedIterator + ExactSizeIterator,
         {
-            // The function body below only uses `self.a/b.len()` and `self.a/b.next_back()`
+            // The function body below only uses `self.a/b.len()` and `self.a/b.next/nth_back()`
             // and doesn’t call `next_back` too often, so this implementation is safe in
             // the `TrustedRandomAccessNoCoerce` specialization
 
             let a_sz = self.a.len();
             let b_sz = self.b.len();
-            if a_sz != b_sz {
-                // Adjust a, b to equal length
-                if a_sz > b_sz {
-                    for _ in 0..a_sz - b_sz {
-                        self.a.next_back();
-                    }
-                } else {
-                    for _ in 0..b_sz - a_sz {
-                        self.b.next_back();
-                    }
-                }
-            }
-            match (self.a.next_back(), self.b.next_back()) {
+            let next_opts = if a_sz == b_sz {
+                (self.a.next_back(), self.b.next_back())
+            } else if a_sz > b_sz {
+                (self.a.nth_back(a_sz - b_sz), self.b.next_back())
+            } else {
+                // Equalize lengths before calling next_back,
+                // so iterators are called in correct order
+                self.b.nth_back(b_sz - a_sz - 1);
+                (self.a.next_back(), self.b.next_back())
+            };
```

Comment:
I think this would be cleaner if you used `match a_sz.cmp(b_sz) { Ordering::... => ... }`